app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt

```diff
@@ -5,65 +5,111 @@ import org.p2p.solanaj.kits.transaction.TransactionDetails
-
-private const val PAGE_LIMIT = 20
+import org.p2p.wallet.user.interactor.UserInteractor
+import timber.log.Timber
 
 class HistoryInteractor(
-    private val rpcSignatureRepository: RpcSignatureRepository,
     private val rpcAccountRepository: RpcAccountRepository,
-    private val transactionsRemoteRepository: TransactionDetailsRemoteRepository,
     private val transactionsLocalRepository: TransactionDetailsLocalRepository,
+    private val transactionsRemoteRepository: TransactionDetailsRemoteRepository,
     private val tokenKeyProvider: TokenKeyProvider,
-    private val historyTransactionMapper: HistoryTransactionMapper
+    private val historyTransactionMapper: HistoryTransactionMapper,
+    private val rpcSignatureRepository: RpcSignatureRepository,
+    private val userInteractor: UserInteractor
 ) {
+    private val allSignatures = mutableListOf<RpcTransactionSignature>()
+    private lateinit var multipleStreamSource: HistoryStreamSource
+    private val accountsStreamSources = HashMap<String, HistoryStreamSource>()
+    private val historyStreamSources = mutableListOf<HistoryStreamSource>()
 
-    suspend fun loadSignaturesForAddress(
-        tokenPublicKey: String,
-        before: String? = null
-    ): List<RpcTransactionSignature> {
-
-        val signatures = rpcSignatureRepository.getConfirmedSignaturesForAddress(
-            userAccountAddress = tokenPublicKey.toPublicKey(),
-            before = before,
-            limit = PAGE_LIMIT
+    private suspend fun initStreamSources() {
+        historyStreamSources.addAll(
+            userInteractor.getUserTokens().map {
+                val accountStreamSource = AccountStreamSource(it.publicKey, rpcSignatureRepository)
+                accountsStreamSources[it.publicKey] = accountStreamSource
+                accountStreamSource
+            }
         )
-        return signatures.map { RpcTransactionSignature(it.signature, it.confirmationStatus) }
+        multipleStreamSource = MultipleStreamSource(historyStreamSources)
+    }
+
+    suspend fun loadTransactions(): List<HistoryTransaction> {
+        if (historyStreamSources.isEmpty()) {
+            initStreamSources()
+        }
+        val signatures = loadAllSignatures()
+        allSignatures.addAll(signatures)
+        return loadTransactions(signatures)
+    }
+
+    suspend fun loadTransactions(account: String): List<HistoryTransaction> {
+        if (historyStreamSources.isEmpty()) {
+            initStreamSources()
+        }
+        val signatures = loadSignaturesForAccount(account)
+        return loadTransactions(signatures)
+    }
+
+    private suspend fun loadAllSignatures(): MutableList<RpcTransactionSignature> {
+        val transactionsSignatures = mutableListOf<RpcTransactionSignature>()
+        while (true) {
+            val firstItem = multipleStreamSource.currentItem() ?: break
+            val time = (firstItem.streamSource?.blockTime ?: -1) - (60 * 60 * 24)
+
+            while (true) {
```

Comment:
Please add comments to the class, how the history works. 
And add todo about the refactoring